ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp

```diff
@@ -325,6 +325,69 @@ tuple<uint32_t, uint32_t> get_matmul_subblock_params(
     return {1, 1};
 }
 
+
+tt::operations::primary::MatmulProgramConfig get_reuse_matmul_program_config(
+    const Tensor& input_tensor_a,
+    const Tensor& input_tensor_b,
+    const MemoryConfig& output_mem_config,
+    const bool fp32_dest_acc_en,
+    const CoreCoord& grid_size) {
+    // TODO: Need a better criteria for BMMs and MatmulMultiCoreReuseProgramConfig
+    TT_FATAL(input_tensor_a.memory_config().memory_layout != TensorMemoryLayout::WIDTH_SHARDED);
+
+    bool per_core_N_equals_subblock_w_constraint = false;
+    if (output_mem_config.is_sharded()) {
+        TT_FATAL(input_tensor_a.memory_config().buffer_type == output_mem_config.buffer_type);
+        TT_FATAL(input_tensor_a.memory_config().memory_layout == output_mem_config.memory_layout);
+        per_core_N_equals_subblock_w_constraint = true;
+    }
+
+    uint32_t M = input_tensor_a.volume() / input_tensor_a.get_legacy_shape()[-1] / TILE_HEIGHT;
+    uint32_t K = input_tensor_a.get_legacy_shape()[-1] / TILE_WIDTH;
+    uint32_t N = input_tensor_b.get_legacy_shape()[-1] / TILE_WIDTH;
+
+    auto a_shard_spec = input_tensor_a.shard_spec();
+    uint32_t per_core_M;
+    uint32_t per_core_N = N;
+    uint32_t in0_block_w;
+    if (a_shard_spec.has_value()) {
+        auto in0_shard_shape = input_tensor_a.shard_spec().value().shape;
+        per_core_M = in0_shard_shape[0] / TILE_HEIGHT;
+        in0_block_w = in0_shard_shape[1] / TILE_WIDTH;
+    } else {
+        per_core_M = M;
+        in0_block_w = 1;
```

Comment:
why is in0_block_w set to 1? or in this case, we simply don't care about performance.

Reply:
I'm basing this on another line for the same uOP where the inputs are interleaved:
```
k_tiles_per_core = 1;  // TODO(arakhmati): Can it be more than 1 without running out of memory?
```
What would be a good setting?
`(K % 2) == 0 ? 2 : 1` ?

I'm open to changing per_core_M as well, although I'm not sure what that should be.

Reply:
per_core_M=M basically means it uses a single core or 1D matmul, this is for the case when user don't provide a grid right? if user don't provide a grid, then are we falling back to single core?

(K % 2) == 0 ? 2 : 1 if it is single core, maybe 1 is better since there won't be memory issue

Reply:
re: per_core_M=M, that's what the existing functionality looks like to me, yes. I'm wondering, do you have an opinion about something better that we can do? If not, that can be looked at in a different PR.

re: (K... ok.